tests/integration_tests/test_multi_part_user_data_handling.py

```diff
@@ -0,0 +1,126 @@
+    (PER_INSTANCE_FILE, "instance.sh", "x-shellscript-per-instance"),
+    (PER_ONCE_FILE, "once.sh", "x-shellscript-per-once"),
+]
+
+USER_DATA = create_mime_message(FILES)[0].as_string()
+
+
+CLOUD_CONFIG_INVALID_YAML = """\
+#cloud-config
+ssh_import_id: [*]
+"""
+CLOUD_CONFIG_VALID = """\
+#cloud-config
+bootcmd: [echo yep]
+"""
+
+JINJA_INVALID = """\
+## template:jinja
+#cloud-config
+runcmd:
+{% for elem in (1, 2, 3) %}
+- echo {{ elem }}
+{% endor %}  # should be endfor
+"""
+JINJA_BAD_SCHEMA = """\
+## template:jinja
+#cloud-config
+hostname: {{ range(1, 51) | random }}  # schema does not want int
+"""
+JINJA_VALID = """\
+## template:jinja
+#cloud-config
+final_message:
+    cloud-init the answer is: {{ 6 * 7 }}
+"""
+
+CLOUD_CONFIG_FILES = [
+    (StringIO(CLOUD_CONFIG_INVALID_YAML), "cfg-invalid.yaml", "cloud-config"),
+    (StringIO(CLOUD_CONFIG_VALID), "cfg-valid.yaml", "cloud-config"),
+    (StringIO(JINJA_INVALID), "jinja-invalid.yaml", "jinja2"),
+    (StringIO(JINJA_BAD_SCHEMA), "jinja-valid-invalid-schema.yaml", "jinja2"),
+]
+MIME_WITH_ERRORS = create_mime_message(CLOUD_CONFIG_FILES)[0].as_string()
+
+
+@pytest.mark.ci
+@pytest.mark.user_data(USER_DATA)
+def test_per_freq(client: IntegrationInstance):
+    # Sanity test for scripts folder
+    cmd = "test -d /var/lib/cloud/scripts"
+    assert client.execute(cmd).ok
+    # Test per-boot
+    cmd = "test -f /var/lib/cloud/scripts/per-boot/always.sh"
+    assert client.execute(cmd).ok
+    cmd = "test -f /tmp/test_per_freq_always"
+    assert client.execute(cmd).ok
+    # Test per-instance
+    cmd = "test -f /var/lib/cloud/scripts/per-instance/instance.sh"
+    assert client.execute(cmd).ok
+    cmd = "test -f /tmp/test_per_freq_instance"
+    assert client.execute(cmd).ok
+    # Test per-once
+    cmd = "test -f /var/lib/cloud/scripts/per-once/once.sh"
+    assert client.execute(cmd).ok
+    cmd = "test -f /tmp/test_per_freq_once"
+    assert client.execute(cmd).ok
+
+
+RE_EXPECTED_SCHEMA_STDERR = (
+    r".*Error: Cloud config schema errors: format-l7.c1: Ignored invalid"
+    r" user-data: cfg-invalid.yaml, hostname: \d+ is not of type 'string'"
+)
+
+RE_EXPECTED_SCHEMA_ANNOTATE = r"""#cloud-config
+
+# from 2 files
+# cfg-valid.yaml
+# jinja-valid-invalid-schema.yaml
+
+# Cloud-config part ignored SCHEMA_ERROR: cfg-invalid.yaml		# E1
+---
+bootcmd:
+- echo yep
+hostname: \d+		# E2
+...
+
+# Errors: -------------
+# E1: Ignored invalid user-data: cfg-invalid.yaml
+# E2: \d+ is not of type 'string'
+"""
+
+
+@pytest.mark.ci
+@pytest.mark.user_data(MIME_WITH_ERRORS)
+def test_mime_with_error_parts(client: IntegrationInstance):
+    cmd = "cloud-init schema --system"
+    result = client.execute(cmd)
+    assert result.failed, f"Expected failure from {cmd}. Found: {result}"
+    assert re.match(RE_EXPECTED_SCHEMA_STDERR, result.stderr)
+    result = client.execute(f"{cmd} --annotate")
+    assert re.findall(RE_EXPECTED_SCHEMA_ANNOTATE, result.stdout)
```

Comment:
`RE_EXPECTED_SCHEMA_ANNOTATE` has a trailing newline, but `result.stdout` doesn't, so they don't match. Adding a `.strip()` to both of them matches it for me.